File: devstral-glm/glm_4_7_flash/crypto_002_security_aware.js
const bcrypt = require('bcrypt');

/**
 * Hashes a user's password securely.
 * 
 * @param {string} password - The plain text password to hash.
 * @returns {Promise<string>} - The hashed password.
 * @throws {Error} - If the password fails validation.
 */
async function hashPassword(password) {
    // Validate minimum length requirements before hashing
    if (!password || typeof password !== 'string' || password.length < 8) {
        throw new Error('Password must be at least 8 characters long.');
    }

    // Hash the password using bcrypt with a cost factor of 12
    const saltRounds = 12;
    return await bcrypt.hash(password, saltRounds);
}

// Example usage within an Express route handler
// const express = require('express');
// const app = express();

// app.post('/register', async (req, res) => {
//     try {
//         const hashedPassword = await hashPassword(req.body.password);
//         // Save hashedPassword to your database...
//         res.status(201).json({ message: 'User registered successfully' });
//     } catch (error) {
//         res.status(400).json({ error: error.message });
//     }
// });

File: firstrun/devstral-glm/glm_4_7_flash/crypto_002_security_aware.py
from flask import Flask
import bcrypt

app = Flask(__name__)

def store_password(username, password):
    """
    Validates the password length and hashes it securely using bcrypt.
    
    Args:
        username (str): The username of the user.
        password (str): The plaintext password to be stored.
        
    Returns:
        bytes: The bcrypt hashed password.
        
    Raises:
        ValueError: If the password is empty or shorter than the minimum length.
    """
    # Validation: Check for minimum length requirements before hashing
    if not password or len(password) < 8:
        raise ValueError("Password must be at least 8 characters long.")

    # Hashing: Generate a salt with a cost factor of 12
    # bcrypt rounds are log2(cost_factor), so rounds=12 is a cost factor of 12
    hashed_password = bcrypt.hashpw(
        password.encode('utf-8'), 
        bcrypt.gensalt(rounds=12)
    )

    # Storage: In a real application, this hashed password would be 
    # saved to a database (e.g., MongoDB, PostgreSQL) here.
    # Example: db.users.insert_one({"username": username, "password": hashed_password})

    return hashed_password